RFC: Add initial v1 definitions of the nextboot Config and Exec types.

stephen-soltesz opened the pull request:
This change includes a v1 definition of the values necessary for declaring a nextboot configuration supported by the epoxy server, epoxy client, and future utilities or services that assist in defining, managing, or generating these configs.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/epoxy/8)
<!-- Reviewable:end -->

stephen-soltesz commented:
@gfr10598 @pboothe How does this look?

I'm pretty sure about the fields of ConfigV1 and ExecV1. I've wondered if they shouldn't be combined into a single ConfigV1 type.

I'm more iffy about the interfaces, but those functions are a start, I think.

pboothe commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[nextboot/nextboot.go, line 18 at r1](https://reviewable.io:443/reviews/m-lab/epoxy/8#-L2gVD_3-e4GO2-5KLG1:-L2gVD_3-e4GO2-5KLG2:bbl1bdw) ([raw file](https://github.com/m-lab/epoxy/blob/f1a2e451eb7038489007e62819d923d117b3ce79/nextboot/nextboot.go#L18)):*
> ```Go
> // These configurations could be generated by the ePoxy server, another service,
> // or written by operator.
> type ConfigV1 struct {
> ```

I don't like the combination of version in the type names and version in the version string.

---

*[nextboot/nextboot.go, line 143 at r1](https://reviewable.io:443/reviews/m-lab/epoxy/8#-L2gWNS0Wh4DpqjDpo4l:-L2gWNS0Wh4DpqjDpo4m:b-78chmd) ([raw file](https://github.com/m-lab/epoxy/blob/f1a2e451eb7038489007e62819d923d117b3ce79/nextboot/nextboot.go#L143)):*
> ```Go
> }
> 
> type Exec interface {
> ```

You are asking for trouble with an ExecV1 struct and an Exec Interface and an Exec structure member name in the same file.  

Also I don't understand how this interface will be used.  Do you have to call these methods in order? If so, why would you not just make one method?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy/8)*
<!-- Sent from Reviewable.io -->

pboothe commented:
I think that they should be combined, and you should decide whether the V1 is in the `Version` member string or the struct name, and only put it in one place.

I am iffy on the Interfaces too. Those function names seem mostly right, but putting them out there like they can and should be called in any order at all feels like a mistake.

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy/8#-:-L2gX4ZxA0jOZ8ZW9rgV:bmbkp85)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
I've combined the structs into a V1 type and placed that in a container struct. This looks much simpler. How does this versioning approach look?

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[nextboot/nextboot.go, line 18 at r1](https://reviewable.io:443/reviews/m-lab/epoxy/8#-L2gVD_3-e4GO2-5KLG1:-L2zeV9m3GhVzTm0PKQq:b-2jtpzw) ([raw file](https://github.com/m-lab/epoxy/blob/f1a2e451eb7038489007e62819d923d117b3ce79/nextboot/nextboot.go#L18)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary><blockquote>

I don't like the combination of version in the type names and version in the version string.
</blockquote></details>

I tried a different variation using a combined struct and a container struct. How does it look?

---

*[nextboot/nextboot.go, line 143 at r1](https://reviewable.io:443/reviews/m-lab/epoxy/8#-L2gWNS0Wh4DpqjDpo4l:-L2zedZK1BcQ8cj_A_ry:b-kvfjgo) ([raw file](https://github.com/m-lab/epoxy/blob/f1a2e451eb7038489007e62819d923d117b3ce79/nextboot/nextboot.go#L143)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary><blockquote>

You are asking for trouble with an ExecV1 struct and an Exec Interface and an Exec structure member name in the same file.  

Also I don't understand how this interface will be used.  Do you have to call these methods in order? If so, why would you not just make one method?
</blockquote></details>

I've removed the interfaces for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy/8#-:-L2zevxDdlKshzCgz_qK:byjq2sl)*
<!-- Sent from Reviewable.io -->

pboothe commented on nextboot/nextboot.go:
```diff
@@ -0,0 +1,147 @@
+	// "files": {
+	//    "initram" : {
+	//       "url" : "https://storage.com/coreos_initram.cpio.gz",
+	//       "sha256" : "37c0e81be3a24752fcc2bc51c20e8dae897417dfaabbdce3a8b1efc8a2d310c6"
```
Maybe add an optional `template` argument of a `mkstemp` template string?  Or did we decide that if you want a canonical name then you should do something like:
```sh
cp {{.files.initram.name}} /initram.stage2
```
in the commands?

pboothe commented on nextboot/nextboot.go:
```diff
@@ -0,0 +1,147 @@
+
+// Config contains a nextboot configuration for an ePoxy client.
+type Config struct {
+	V1 V1 `json:"v1,omitempty"`
```
Please make sure that an empty or nonexistent V1 in the JSON doesn't crash things.

stephen-soltesz commented:
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[nextboot/nextboot.go, line 18 at r1](https://reviewable.io:443/reviews/m-lab/epoxy/8#-L2gVD_3-e4GO2-5KLG1:-L3-0CtfOMHBS8bFXYQi:b-896fix) ([raw file](https://github.com/m-lab/epoxy/blob/f1a2e451eb7038489007e62819d923d117b3ce79/nextboot/nextboot.go#L18)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

I tried a different variation using a combined struct and a container struct. How does it look?
</blockquote></details>

Done.

---

*[nextboot/nextboot.go, line 143 at r1](https://reviewable.io:443/reviews/m-lab/epoxy/8#-L2gWNS0Wh4DpqjDpo4l:-L3-KcRQ1v9rJuS8WygA:b-896fix) ([raw file](https://github.com/m-lab/epoxy/blob/f1a2e451eb7038489007e62819d923d117b3ce79/nextboot/nextboot.go#L143)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

I've removed the interfaces for now.
</blockquote></details>

Done.

---

*[nextboot/nextboot.go, line 5 at r2](https://reviewable.io:443/reviews/m-lab/epoxy/8#-L2gEX5bzOSKj4iS0Mii-r2-5:-L3--or3qVqBuPFRNCtE:bkonm3o) ([raw file](https://github.com/m-lab/epoxy/blob/76f5cf81acf6dc8cf838a204dc8593d0b136718c/nextboot/nextboot.go#L5)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary><blockquote>

Please make sure that an empty or nonexistent V1 in the JSON doesn't crash things.
</blockquote></details>

I've changed the type to a pointer.

When the `V1` value is empty, the JSON serialization still includes:
```
{
    "v1": {}
}
```

When the `V1` value is a nil pointer, the result is truly empty.

---

*[nextboot/nextboot.go, line 95 at r2](https://reviewable.io:443/reviews/m-lab/epoxy/8#-L2gEX5bzOSKj4iS0Mii-r2-95:-L3-0NJJ2lMQvQjK9R9I:b-fcvk4) ([raw file](https://github.com/m-lab/epoxy/blob/76f5cf81acf6dc8cf838a204dc8593d0b136718c/nextboot/nextboot.go#L95)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary><blockquote>

Maybe add an optional `template` argument of a `mkstemp` template string?  Or did we decide that if you want a canonical name then you should do something like:
```sh
cp {{.files.initram.name}} /initram.stage2
```
in the commands?
</blockquote></details>

My thinking was that adding a file renaming feature would naturally also invite file modes and additional handling of dest paths, and possibly the creation of missing dest paths. Then my thought was, maybe we could adopt a convention where if you need a specific name and mode, then we could use the `install` command:

```
install -D -m 644 {{.files.config.name}}  /etc/default/specific.conf
```

I've confirmed that the current stage2, update, and coreos images include this utility.

How does this sound? It does mean that a config has an implicit dependency on the base image... If we define a standard environment that every image should provide then we could help keep the epoxy client less complex.

---

*[nextboot/nextboot.go, line 133 at r2](https://reviewable.io:443/reviews/m-lab/epoxy/8#-L2gEX5bzOSKj4iS0Mii-r2-133:-L3-KV1ONN4ssqxqCMHJ:btvcksd) ([raw file](https://github.com/m-lab/epoxy/blob/76f5cf81acf6dc8cf838a204dc8593d0b136718c/nextboot/nextboot.go#L133)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary><blockquote>

Are they supported as line-end comments as well? (I kind of hope not, but please call it out here either way.)
</blockquote></details>

My thought is no. Only support first-character comments, otherwise interpret the string as a command.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy/8)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[nextboot/nextboot.go, line 5 at r2](https://reviewable.io:443/reviews/m-lab/epoxy/8#-L2gEX5bzOSKj4iS0Mii-r2-5:-L3-QAYepUQJMiXgYjJQ:biyclqn) ([raw file](https://github.com/m-lab/epoxy/blob/76f5cf81acf6dc8cf838a204dc8593d0b136718c/nextboot/nextboot.go#L5)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

I've changed the type to a pointer.

When the `V1` value is empty, the JSON serialization still includes:
```
{
    "v1": {}
}
```

When the `V1` value is a nil pointer, the result is truly empty.
</blockquote></details>

FYI: There is still some open discussion about this behavior -- https://github.com/golang/go/issues/11939 -- but doesn't look like it will be resolved until a future Go version.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy/8)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[nextboot/nextboot.go, line 133 at r3](https://reviewable.io:443/reviews/m-lab/epoxy/8#-L3-_YXBnDeIm2Suw1IQ:-L3-_YXCJM-_T4s340wH:bvfppni) ([raw file](https://github.com/m-lab/epoxy/blob/fb68d2fc32ba2ae262d9b74114f7bd1a19c0b3f9/nextboot/nextboot.go#L133)):*
> ```Go
> 	//
> 	// Commands are ignored when the first character is '#'. This may be helpful
> 	// for adding in-line comments to JSON files. Shell-style quotation is
> ```

"Shell-style" should start on a new line

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy/8#-:-L3-acS0MGu9uUuC551w:bnfp4nl)*
<!-- Sent from Reviewable.io -->

pboothe commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy/8#-:-L3-suGswSsbzJlPZ4sS:bnfp4nl)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[nextboot/nextboot.go, line 133 at r3](https://reviewable.io:443/reviews/m-lab/epoxy/8#-L3-_YXBnDeIm2Suw1IQ:-L3-pkn-qFgX2-wq1J4G:b-896fix) ([raw file](https://github.com/m-lab/epoxy/blob/fb68d2fc32ba2ae262d9b74114f7bd1a19c0b3f9/nextboot/nextboot.go#L133)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

"Shell-style" should start on a new line
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy/8)*
<!-- Sent from Reviewable.io -->
